uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js

```diff
@@ -248,22 +250,52 @@ define([
       };
 
       /**
-       * Flattens statuses from object to array of objects. This is used to
-       * populate the dropdown with array of statuses.
-       * Also it checks if given status is available to manager. If manager applies leave
-       * on behalf of staff then cancelled is also removed from her list of available statuses.
+       * Returns an array of statuses depending on the previous status value
+       * This is used to populate the dropdown with array of statuses.
        *
        * @return {Array}
        */
+
       this.getStatuses = function () {
-        return _.reject(this.requestStatuses, function (status) {
-          var canRemoveStatus = (status.name === sharedSettings.statusNames.adminApproved || status.name === sharedSettings.statusNames.awaitingApproval);
+        if (!this.request || angular.equals({}, vm.requestStatuses)) {
+          return [];
+        }
 
-          return this.canManage ? (canRemoveStatus || status.name === 'cancelled') : canRemoveStatus;
-        }.bind(this));
+        if (!this.request.status_id) {
+          return getAvailableStatusesForStatusName('none');
+        }
+
+        return getAvailableStatusesForCurrentStatus();
       };
 
       /**
+       * Helper functions to get available statuses depending on the
+       * current request status value.
+       *
+       * @return {Array}
+       */
+      function getAvailableStatusesForCurrentStatus () {
+        var currentStatus = vm.getStatusFromValue(vm.request.status_id);
+
+        return getAvailableStatusesForStatusName(currentStatus.name);
+      }
+
+      /**
+       * Helper function that returns an array of the statuses available
+       * for a specific status name.
+       *
+       * @return {Array}
+       */
+      function getAvailableStatusesForStatusName (statusName) {
+        return _.filter(vm.requestStatuses, function (nextStatus) {
+          return _.includes(
+            availableStatusesMatrix[ statusName ],
+            nextStatus.name
```

Comment:
I'll refactor this function since I think I can use `availableStatusesMatrix` directly. 

nextStatus.name is used here to see if the status is part of the statuses matrix which means it's a valid transition status. Now that I understand the code better I think it's safe to return availableStatuses directly or using a map.